fix: (.NET) Improve type conversion, add parse_value helper, and enhance null handling

Fellmonkey opened the pull request:
<!--
Thank you for sending the PR! We appreciate you spending the time to work on these changes.

Help us understand your motivation by explaining why you decided to make this change.

You can learn more about contributing to appwrite here: https://github.com/appwrite/appwrite/blob/master/CONTRIBUTING.md

Happy contributing!

-->

## What does this PR do?

1. Adds a new `parse_value` Twig function to centralize type conversion logic
2. Refactors the `ObjectToInferredTypesConverter` for better JSON parsing
3. Improves null handling in the Client and Models
4. Replaces `ConvertToList<T>` with `ToEnumerable` for more flexible type handling
5. Fixes namespace references (Appwrite → dynamic spec.title)

## Related PRs and Issues
This PR addresses potential runtime crashes and improves the overall reliability of JSON handling in generated .NET SDKs. The changes are particularly important for applications that work with complex nested JSON structures from API responses.

Related to issues with incorrect type mapping, JsonElement leakage, and runtime errors during deserialization of complex/nested JSON structures.
#[73](https://github.com/appwrite/sdk-for-dotnet/issues/73)


### Have you read the [Contributing Guidelines on issues](https://github.com/appwrite/appwrite/blob/master/CONTRIBUTING.md)?

YES

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
  - Added a general enumerable conversion helper and an exception constructor that accepts an inner exception.
  - Generator now emits unified value-parsing helpers for model mapping and serialization.

- **Bug Fixes**
  - JSON deserialization returns nullable results and explicitly handles null/undefined; improved type inference (dates, numbers, strings, booleans).
  - Request preparation now skips null parameters to avoid processing errors.

- **Refactor**
  - Models and parsing logic reorganized for consistent naming and unified value conversion.

- **Style**
  - Minor formatting and namespace templating updates.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

Fellmonkey committed:
Refactor object inference in JSON converter

Replaces switch on JsonTokenType with a recursive method using JsonElement.ValueKind for more robust and accurate type inference. This improves handling of nested objects and arrays, and unifies the logic for converting JSON values to .NET types.

Fellmonkey committed:
Refactor model deserialization logic in C# template

Simplifies and standardizes the deserialization of model properties from dictionaries, removing special handling for JsonElement and streamlining array and primitive type conversions. This improves code readability and maintainability in generated model classes.

Fellmonkey committed:
Handle optional properties in model From method

Updated the From method in the model template to check for the existence of optional properties in the input map before assigning values. This prevents errors when optional properties are missing from the input dictionary. (for examle in model: User, :-/ )

Fellmonkey committed:
synchronization with the Unity template

ChiragAgg5k commented on templates/dotnet/Package/Converters/ObjectToInferredTypesConverter.cs.twig:
```diff
@@ -41,4 +69,4 @@ namespace {{ spec.title | caseUcfirst }}.Converters
         }
     }
-}
+}
```
can we add back the linebreak at the end?

ChiragAgg5k commented on templates/dotnet/Package/Enums/Enum.cs.twig:
```diff
@@ -16,4 +16,4 @@ namespace {{ spec.title | caseUcfirst }}.Enums
         {%~ endfor %}
     }
-}
+}
```
same here

Fellmonkey committed:
Refactor model parsing for nullable and array properties

Improves the From() method in Model.cs.twig to handle nullable and array properties more robustly, using helper macros for parsing arrays and sub-schemas. This change ensures correct handling of optional fields and type conversions, reducing runtime errors and improving code maintainability. Also removes an unnecessary blank line in ServiceTemplate.cs.twig.

Fellmonkey committed:
Skip null parameters in request parameter loop

Fields with null values in multipart are now omitted (so they don't turn into empty strings).

Fellmonkey commented:
Fields with null values in multipart are now omitted (so they don't turn into empty strings).
<img width="875" height="272" alt="image" src="https://github.com/user-attachments/assets/e7fb909d-6d4a-4df9-b815-779b93fc83a2" />
<img width="322" height="110" alt="{861587C0-66F7-4B57-BD7B-3DF4766EF3D3}" src="https://github.com/user-attachments/assets/dd3b8fc2-9e6f-4214-98ae-4df4adeedd65" />

Fellmonkey committed:
Refactor model class name generation in template

Fellmonkey committed:
Merge branch 'master' into improve-json-serialization

Fellmonkey committed:
Add parse_value Twig function for DotNet models

Introduces a new parse_value Twig function in DotNet.php to centralize and simplify value parsing logic for model properties. Updates Model.cs.twig to use this function, reducing template complexity and improving maintainability.

Fellmonkey committed:
make generated array mappings null-safe

Remove null-forgiving operator (!) from optional array mappings and use null-safe casting to preserve null vs empty semantics in generated models.

Fellmonkey committed:
lint

Fellmonkey committed:
Import Enums namespace conditionally in model template

Adds conditional import of the Enums namespace in the Model.cs.twig template only when the model definition contains enum properties. This prevents unnecessary imports and improves template clarity.

ChiragAgg5k commented:
@Fellmonkey waiting on jake's review, will merge then 👍🏻

Fellmonkey commented:
> Hello @ChiragAgg5k, I made some minor adjustments, wrote some tests, and noticed that two arguments that are required in the code in the API are actually missing. It would be best to make them null-safe or see why the API is not returning them. <img alt="{7CF5D045-1E68-46F6-A2E3-486785466DD8}" width="516" height="253" src="https://private-user-images.githubusercontent.com/90258055/494099110-565601ee-e2ed-4722-8265-9b7dae4aec9c.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTk1Njc0MjgsIm5iZiI6MTc1OTU2NzEyOCwicGF0aCI6Ii85MDI1ODA1NS80OTQwOTkxMTAtNTY1NjAxZWUtZTJlZC00NzIyLTgyNjUtOWI3ZGFlNGFlYzljLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTEwMDQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUxMDA0VDA4Mzg0OFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWY4ZGMwNDVkNDNjMDAwMTZjZDBjNTZmNzUwNTljMTU5NzhmMDIxNjZkYWU5ODkxZDAxM2M2MTRiMjQ5YjEzOWYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.6f9mgP3jdc8ZC80M7bP1Q0uMo7R56uNqvfOCLEeOJaw">
> 
> <img alt="image" width="822" height="174" src="https://private-user-images.githubusercontent.com/90258055/494099007-4916752b-7ad4-4d15-a4eb-459c9edb8457.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTk1Njc0MjgsIm5iZiI6MTc1OTU2NzEyOCwicGF0aCI6Ii85MDI1ODA1NS80OTQwOTkwMDctNDkxNjc1MmItN2FkNC00ZDE1LWE0ZWItNDU5YzllZGI4NDU3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTEwMDQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUxMDA0VDA4Mzg0OFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTUyMDg4Zjk3Yjg4NjEzZTZiYmRlYzliNTY0YjllZWRjODQxMjY1MjFhZWZkOTE0ZjQyM2NlMzcwYzU1MDdhMWImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.NZCugpadXHGbSIlWNh9VDh3gNEeRMCFMrd0Vi0JkMSQ"> <img alt="{CB367DEA-584F-4434-BE90-4C4C135E57E3}" width="568" height="159" src="https://private-user-images.githubusercontent.com/90258055/494099936-b06dcf68-4018-4ef1-9745-c7f51921622d.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTk1Njc0MjgsIm5iZiI6MTc1OTU2NzEyOCwicGF0aCI6Ii85MDI1ODA1NS80OTQwOTk5MzYtYjA2ZGNmNjgtNDAxOC00ZWYxLTk3NDUtYzdmNTE5MjE2MjJkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTEwMDQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUxMDA0VDA4Mzg0OFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTYwNjM0NDMyODdiNzg3Yzg5ZTU4YmJkMGQ2NjhlOWRhZjE3NzE5ODNjZmRkNzVjNDRlNDg1NWNhOWZjYjAxZmQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.XfyzhWxxIZmXNf8p3Fe9DGH67v2fCzs1Dl4aEJL1JlQ"> <img alt="{3BB6BE36-2545-4C44-8963-86834212B236}" width="849" height="295" src="https://private-user-images.githubusercontent.com/90258055/494100003-25bcd5fa-2ef8-4346-ad27-2ab59dc3c5ba.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTk1Njc0MjgsIm5iZiI6MTc1OTU2NzEyOCwicGF0aCI6Ii85MDI1ODA1NS80OTQxMDAwMDMtMjViY2Q1ZmEtMmVmOC00MzQ2LWFkMjctMmFiNTlkYzNjNWJhLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTEwMDQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUxMDA0VDA4Mzg0OFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTI4ZjAzZDFmMjUxNDg5YWUyMWZmZGI5MzYwMzA5ZjU5OWJhNGMwZDlhYTJhMzU4YmQwNGZiODkxZTllYTIwMzUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.WQoR6tUIrkaOYa9XwBjSblO1J7gngwMTuWZeXxgusZk">


It's strange here.

I checked GetDB and GetCache server definitely doesn't return any keys to the HealthStatus model.

We need to either edit the model specifications or see why the sever isn't returning this data.

ChiragAgg5k commented:
@Fellmonkey oh sorry missed your comment there, best to check in the appwrite codebase on why those values are not being returned, or if the model requires some adjustments

i did some testing myself and it seems to work just fine for me:

<img width="1143" height="925" alt="Screenshot 2025-10-04 at 9 17 05 PM" src="https://github.com/user-attachments/assets/0388e149-ad61-4cf2-990d-73a24ec0d063" />

ChiragAgg5k committed:
Merge branch 'master' into improve-json-serialization

Fellmonkey commented:
@ChiragAgg5k It really works
<img width="584" height="484" alt="{C7065359-2D91-4F62-9E56-8C6AA1C151C7}" src="https://github.com/user-attachments/assets/6dcc0468-dcf1-4d0c-8ed8-2f36cd1e1034" />

But I don't understand what the problem is with the SDK; the Dart server has the same problem.

<img width="727" height="977" alt="{95741ADA-17E0-429D-A83D-03E51F7393FB}" src="https://github.com/user-attachments/assets/b9139ef8-3bd6-4a33-a2ea-5af7c59d179a" />

I made some changes in Dart, because otherwise an error would occur: type 'Null' is not a subtype of type 'int'
<img width="696" height="570" alt="{834C72F4-FC45-49B5-B9A9-21D8C097CEE1}" src="https://github.com/user-attachments/assets/976daf92-c444-4c68-bf87-082fdad9fcca" />

ChiragAgg5k commented:
@Fellmonkey interesting it doesnt work in dart, would be great to figure out why. can you please create a seperate issue for tracking this 🙏

Fellmonkey committed:
Refactor array handling in DotNet code generation

Introduces a ToEnumerable extension method to unify array and enumerable conversions in generated .NET code. Updates code generation logic to use ToEnumerable for array properties, simplifying and improving type safety. Also adds necessary using statement for Extensions in generated model files.

Fellmonkey committed:
Merge branch 'master' into improve-json-serialization

Fellmonkey committed:
Merge remote-tracking branch 'upstream/master' into improve-json-serialization

Fellmonkey committed:
Update docblock for getFunctions method

Fellmonkey committed:
Refactor exception and extension handling

Fellmonkey committed:
Refactor .NET model property serialization

Introduce centralized property conversion and mapping helpers for the .NET SDK generator and update model template to use them.

Changes:
- Add Twig filters: propertyAssignment and toMapValue.
- Make getPropertyType optionally return non-fully-qualified names.
- Replace inline parsing logic with new methods: getPropertyName, getResolvedPropertyName, getPropertyAssignment, convertValue, getToMapExpression. These handle enums, sub-schemas, arrays, primitive conversions and null-safety in a unified way.
- Simplify getFunctions() implementation to delegate to the new helpers.
- Update templates/dotnet/Package/Models/Model.cs.twig to use the new filters (propertyAssignment, toMapValue) and clean up From/ToMap generation.

Reason: centralizes and standardizes property (de)serialization logic, improves null handling, supports property overrides, and simplifies the model template.

Fellmonkey commented on src/SDK/Language/DotNet.php:
```diff
@@ -548,39 +555,171 @@ public function getFunctions(): array
+        // String, boolean, arrays — null-safe conversion
+        $expr = $this->convertValue($property, $v, false);
+        return "{$tryGet} ? {$expr} : null";
+    }
```
According to the specification, arrays are always [], and elements are required/non-nullable.

Fellmonkey committed:
Always call ToList() for mapped arrays

Fellmonkey committed:
Merge branch 'master' into improve-json-serialization

Fellmonkey commented on src/SDK/Language/DotNet.php:
```diff
@@ -548,39 +555,171 @@ public function getFunctions(): array
+                return "{$propName}{$nullOp}.Select(it => it.ToMap())" . (!$required ? '?.ToList()' : '');
+            }
+            return "{$propName}{$nullOp}.ToMap()";
+        }
```
fixed in 8fe9521032ad5442bc39649dd98e0d0244661a5b

Fellmonkey committed:
Add missing closing for toMapValue TwigFilter
